src/components/diagrams/diagram-grid-layout.tsx

```diff
@@ -347,26 +411,42 @@ function DiagramGridLayout({ studyUuid, showInSpreadsheet, visible }: Readonly<D
         visible,
     ]);
 
-    const onLoadFromSessionStorage = useCallback((savedLayouts: Layouts | undefined) => {
-        if (savedLayouts) {
+    const onLoadDiagramLayout = useCallback((savedLayouts: Layouts) => {
+        if (Object.keys(savedLayouts).length > 0) {
             const savedLayoutsEntries = Object.entries(savedLayouts);
             const newLayoutsEntries = savedLayoutsEntries.map(([breakpoint, breakpoint_layouts]) => {
                 const updatedLayouts = [...breakpoint_layouts];
                 updatedLayouts.unshift(AdderCard);
-                return [breakpoint, updatedLayouts];
+                return [
+                    breakpoint,
+                    updatedLayouts.map((l) => ({
+                        ...l,
+                        w: decodeInfinity(l.w),
+                        h: decodeInfinity(l.h),
+                        x: decodeInfinity(l.x),
+                        y: decodeInfinity(l.y),
+                    })),
+                ];
             });
             setLayouts(Object.fromEntries(newLayoutsEntries));
         } else {
             setLayouts(initialLayouts);
         }
     }, []);
 
+    const decodeInfinity = (value: number) => {
+        if (value === MAX_INT32) {
+            return Infinity;
+        }
+        return value;
+    };
+
     const onAddMapCard = useCallback(() => {
         // TODO setLayouts to add a map card
         setIsMapCardAdded(true);
     }, []);
 
-    useDiagramsGridLayoutSessionStorage({ layouts, onLoadFromSessionStorage });
+    useDiagramsGridLayout({ onLoadDiagramLayout });
```

Comment:
I think all the new code in this `diagram-grid-layout.tsx` should be moved in a custom hook (this one for example : `useDiagramsGridLayout` but it should be renamed something like `useDiagramsGridLayoutPersistence` or something.
then usage here :
```
const [saveGridLayout] = useDiagramsGridLayoutPersistence({ onLoadDiagramLayout });
```

then `encodeInfinity`, `decodeInfinity` are global functions.
You encapsulate `frontendToBackendAppLayout` and other utilities in a custom hook where its role is to save and load gridLayout.
The 'view component' `DiagramGridLayout` doesn't manage this work then.
It was the design with the old custom hooks using session storage.

Reply:
I agree, working on making something cleaner